File: KthSmallestInABST.py
# Time Complexity : O(k + h) where h is the height
# Space Complexity : O(h)

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def __init__(self):
        self.result = None
        self.count = 0

    def kthSmallest(self, root: TreeNode, k: int) -> int:
        self.result = -1
        self.inorder(root, k)
        return self.result

    def inorder(self, root: TreeNode, k):
        # Base
        if root == None:
            return
        # logic
        self.inorder(root.left, k)
        self.count += 1
        if self.count == k:
            self.result = root.val
        self.inorder(root.right, k)

#Iterative Approach

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        stack = []
        while root or stack:
            while root :
                stack.append(root)
                root = root.left
            root = stack.pop()
            k-=1
            if k == 0:
                return root.val
            root = root.right
        return -1

File: kThSmallestInaBst.js
// Time Complexity : O(n)
// Space Complexity : O(h)
// Did this code successfully run on Leetcode : yes

// Further optimization by not traversing the right sub tree if k === 0
var kthSmallest = function (root, k) {
  let kthElement = 0;
  const inorderTrav = (root) => {
    // base
    if (!root) return;

    // Logic
    inorderTrav(root.left);
    k--;
    if (!k) return (kthElement = root.val);
    if (k > 0) inorderTrav(root.right);
  };

  inorderTrav(root);
  return kthElement;
};

// Time Complexity : O(n)
// Space Complexity : O(h)
var kthSmallest = function (root, k) {
  let kthElement = 0;
  const inorderTrav = (root) => {
    // base
    if (!root) return;

    // Logic
    inorderTrav(root.left);
    k--;
    if (!k) return (kthElement = root.val);
    inorderTrav(root.right);
  };

  inorderTrav(root);
  return kthElement;
};

// Time Complexity : O(n)
// Space Complexity : O(h + n): recursive stack and array space
var kthSmallest = function (root, k) {
  const resultsArr = [];
  const inorderTrav = (root) => {
    // base
    if (!root) return;

    // logic
    inorderTrav(root.left);
    resultsArr.push(root.val);
    inorderTrav(root.right);
  };

  inorderTrav(root);
  return resultsArr[k - 1];
};

// var kthSmallest = function(root, k) {
//   const inorderTrav = (root) => {
//       // base
//       if (!root) return;

//       // Logic
//       inorderTrav(root.left)
//       k--
//       if (!k) return kthElement = root.val
//       inorderTrav(root.right)
//   }

//   return inorderTrav(root);
// };
